### DISCUSSION TITLE
Some thoughts around documentation

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [sqlalchemy/sqlalchemy#11386] [Ideas]
Hey @zzzeek,
Been using sqlalchemy since sometime, and it's really really amazing. I don't know how to thank you for this amazing job, really a game changer. This library(toolkit) never disappoint me, and everyday I still learn more about functionalities that I didn't know it existed, and that's a question if there is a way to optimize this? to showcase the toolkit in better way? 
tbh I think you might have received this feedback that document is a bit tough to work with, especially at the very beginning, and especially if there's no convention yet how to use sqlalchemy, because there's multiple way to achieve something, and that might gets very confusing soon.
So I was thinking does it make sense to write a document about the way I use sqlalchemy? but then i'm actually questioning who am I to judge what's the correct way of using it. I believe the document should be more narrative,  explicitly explaining different approaches to solve a problem that an end user might face, and let the user judge the way of using it, because that's why sqlalchemy exists as a toolkit.

now I can imagine that i'd end up rewriting the documentation, which has more impact if it's included in SQLAlchemy, so lemme begin with some principles i already have in mind:
- A beginner tutorial should show the way to enable me start using SQLAlchemy quickly, with actually understanding what's happening, and also give me some insight about how much framework is capable of, but not going too deep into detail. 
- Mkdocs is easier to use, that requires some effort to move toward to that, but I can take short step a time, working on a branch to deliver this.
- An example should have imports, so the end user can copy paste it and see it working
- An example could be some form of test, increasing coverage and covering "use cases", and avoid runtime errors
- Ordering (for pages, not section within a page) should not matter once user read tutorial. In other term, it doesn't mean which page I'm reading first, if i need something I can search for it, and reading the section and that specific page should already give me the full picture.
- [Layers of SQLAlchemy](https://docs.sqlalchemy.org/en/20/intro.html) should have in-dependent document, where user can switch between them on top of page (header). and then, their implementation is inside scope

Of course achieving all these is hard, especially considering the complexity, but I'm confident if i start working it on my free time, I can deliver it someday and the community benefits alot.
What do you think about this whole idea?

**Comment by zzzeek** [comment]
hi Mani -

short answer we aren't looking for new "rewrites" of our docs right now, and we are very happy / embedded with Sphinx / RST which remains industry standard for Python documentation.

Beyond that, the scale of SQLAlchemy's documentation is quite large and realistically a "rewrite" wouldn't be a single person, volunteer job.   You'd need a paid team of writers to even approach that goal.    Have you worked as a writer and do you have any documentation you've written you can show me?

**Reply by ManiMozaffar** [reply]
Hi,

> we are very happy / embedded with Sphinx / RST which remains industry standard for Python documentation.

I haven't looked that much into it, the thing i liked about mkdocs was the hooks and tooling around it. (point 3 and 4)

> Beyond that, the scale of SQLAlchemy's documentation is quite large and realistically a "rewrite" wouldn't be a single person, volunteer job.

Yes I agree. But I believe it doesn't need to change over time, it could take time to rewrite parts, but it could be kept as separate branch even having multiple person working on it.

> Have you worked as a writer and do you have any documentation you've written you can show me?

No I haven't worked as technical writer, but i documented alot stuff (not open source usually). but latest thing i documented which is actually open source is:
https://manimozaffar.github.io/aioclock/

**Comment by CaselIT** [comment]
Hi,

In addition to what mike has mentioned, I think an iterative approach is a better approach for this task.
PRs to improve the docs are always appreciated, if you have something more extended in mind regarding a part opening an issue or discussion first is likely better.
Do you have some candidate pages/sections/part that you think could use an improvement?

> A beginner tutorial should show the way to enable me start using SQLAlchemy quickly, with actually understanding what's happening, and also give me some insight about how much framework is capable of, but not going too deep into detail.

There is already that: https://docs.sqlalchemy.org/en/20/orm/quickstart.html
Want something more in depth? That's the unified tutorial that were re-written for v2: https://docs.sqlalchemy.org/en/20/tutorial/index.html

They can for sure be improved, but I would think that improving these is a more pragmatic approach than to starting over.

> Mkdocs is easier to use, that requires some effort to move toward to that, but I can take short step a time, working on a branch to deliver this.

I'm not sure that's as easy as you envision to migrate over. SQLAlchemy uses several custom sphinx plugins, as automation (like the changelog one) and to aid the docs. For example the tool that links parameters of function/classes is a custom plugin that was developed by mike (see for example https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column.params.deferred_raiseload). I'm not sure if mkdocs has something comparable.

Personally I don't think "easier to use" is reason enough to completely change the documentation build system. While I agree that markdown is easier and nicer to write than rst, it's not like the sphinx rst requires years of study to understand.

**Reply by ManiMozaffar** [reply]
Hi,

> In addition to what mike has mentioned, I think an iterative approach is a better approach for this task.

100% agreed.

> PRs to improve the docs are always appreciated, if you have something more extended in mind regarding a part opening an issue or discussion first is likely better.

The problem is how examples are written in rst. I believe examples in documentation is very important, and that's the reason why "FastAPI" or Pydantic document is considered as a good one, because it has example that you can copy paste and see the result immediately.


> For example the tool that links parameters of function/classes is a custom plugin that was developed by mike (see for example https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column.params.deferred_raiseload). I'm not sure if mkdocs has something comparable

If you're talking about the changelog addition to sqlalchemy where it shows which version was this added, probably not. 
But there is actually way better plugins in mkdocs to show parameters of function/classes with its respective documentation.

> it's not like the sphinx rst requires years of study to understand

The reason I'm trying to promote it is because of these 2 points:
- An example should have imports, so the end user can copy paste it and see it working
- An example could be some form of test, increasing coverage and covering "use cases", and avoid runtime errors
- 
Maintaining example, keeping most of docs in doc string is way nicer.

**Reply by CaselIT** [reply]
> The problem is how examples are written in rst

I don't understand what you mean. Why does the markup matter in how the examples are written?

>  But there is actually way better plugins in mkdocs to show parameters of function/classes with its respective documentation.

Do you have examples?

> The reason I'm trying to promote it is because of these 2 points:
> 
>     * An example should have imports, so the end user can copy paste it and see it working
> 
>     * An example could be some form of test, increasing coverage and covering "use cases", and avoid runtime errors

I really don't understand why the markup format matter when speaking about examples here

> Maintaining example, keeping most of docs in doc string is way nicer.

 That seems like how the code is currently organized. Again I don't understand what would be different

**Reply by ManiMozaffar** [reply]
For instance consider this:
https://github.com/sqlalchemy/sqlalchemy/blob/93cfb49572ac56bc320a09b82285bf8ef8cdff57/lib/sqlalchemy/orm/base.py#L762

The example is here:
https://github.com/sqlalchemy/sqlalchemy/blob/93cfb49572ac56bc320a09b82285bf8ef8cdff57/doc/build/orm/mapping_styles.rst#declarative-mapping

But it should be inside doc string. Check my code:
https://github.com/ManiMozaffar/aioclock/blob/main/aioclock/triggers.py#L168

This is good because:
- There's a mkdocs plugin, that lints these codes for you (i think with black not sure) and allows you to also run them all making sure all works. There's no runtime or import issue
- Having code and document at one place, allow you to easily reflect the use cases and updates it as you go forward
- You don't necessarily need to go to documentation to find out use cases.

**Reply by CaselIT** [reply]
I disagree, the section that you are have linked is not the doc-string of `Mapped`, so that would not be the right place to have that example.

When an example makes sense in the docstring it's in the docstring: https://github.com/sqlalchemy/sqlalchemy/blob/93cfb49572ac56bc320a09b82285bf8ef8cdff57/lib/sqlalchemy/orm/_orm_constructors.py#L749-L796

The plugin that runs the code would be nice to have, but overall there does not seem to be much difference
